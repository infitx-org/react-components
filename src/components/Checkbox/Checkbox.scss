@use "sass:map";
@use "../../scss/base/shadows.scss";
@use "../../scss/base/colors.scss";
@use "../../scss/components/input.scss";
@use "../../scss/themes.scss";
@use "../../scss/components/any.scss";

.rc-checkbox__wrapper {
  @include any.rc-component;

  display: inline-flex;
  align-items: center;
  padding: 3px;
}

.rc-checkbox__label {
  display: inline-block;
  cursor: pointer;
  font-size: 14px;
}

.rc-checkbox__label--disabled {
  color: #999;
  cursor: default;
}

.rc-checkbox {
  appearance: none;
  outline: none;
  box-sizing: initial;
  position: relative;
  flex: 0 0 20px;
  display: inline-block;
  margin: 0;
  margin-right: 5px;
  background: linear-gradient(#fafafa, #f4f4f4);
  box-shadow: 0 0 0 0 rgba(0, 0, 0, 0.13);
  transition: all 0.2s ease-out;
  flex-basis: 22px;
  width: 22px;
  height: 20px;
  border: none;
  border-radius: input.$border-radius;
  border-bottom: 2px solid rgba(0, 0, 0, 0.2);

  &:hover {
    box-shadow: shadows.$size rgba(0, 0, 0, 0.13);
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  &:disabled {
    background: #eee;
    box-shadow: none;
  }

  &:checked {
    &::after {
      content: "";
      box-sizing: content-box;
      box-sizing: initial;
      position: absolute;
      left: 6px;
      top: 10px;
      background: #fff;
      width: 2px;
      height: 2px;
      box-shadow: 2px 0 0 #fff, 4px 0 0 #fff, 4px -2px 0 #fff, 4px -4px 0 #fff,
        4px -8px 0 #fff, 4px -6px 0 #fff;
      animation: animateCheckboxSelectedContent 0.2s ease-out 1 forwards;
      transform: rotate(45deg);
    }
  }
}

.rc-checkbox--no-margin {
  margin-right: 0;
}

.rc-checkbox--round {
  border-radius: 50%;
}

.rc-checkbox--semi-checked {
  border-color: #ddd;
  background: #ddd;

  &::after {
    content: "";
    box-sizing: content-box;
    box-sizing: initial;
    position: absolute;
    left: 6px;
    top: 10px;
    background: #999;
    width: 10px;
    height: 2px;
  }
}

@each $kind in themes.$kinds {
  .rc-checkbox--#{$kind} {
    &:checked {
      border-color: map.get(themes.$colors--dark, $kind);
      background: linear-gradient(
        0deg,
        map.get(themes.$colors--dark, $kind),
        map.get(themes.$colors, $kind)
      );

      $color: map.get(themes.$colors--opposite, $kind);

      &::after {
        background: $color;
        box-shadow: 2px 0 0 $color, 4px 0 0 $color, 4px -2px 0 $color,
          4px -4px 0 $color, 4px -8px 0 $color, 4px -6px 0 $color;
      }
    }

    &:checked:disabled {
      border-color: map.get(themes.$colors--light, $kind);
      background: map.get(themes.$colors--light, $kind);
    }

    &:focus {
      box-shadow: shadows.$size map.get(themes.$shadows, $kind);
    }
  }
}

@keyframes animateCheckboxSelectedContent {
  0% {
    margin-top: 2px;
    opacity: 0;
  }

  100% {
    margin-top: 0;
    opacity: 1;
  }
}
