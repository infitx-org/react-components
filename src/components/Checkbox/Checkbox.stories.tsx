import React from "react";
import Checkbox from "./Checkbox";

export default {
  title: "Components/Checkbox",
  component: Checkbox,
};

const Template = (args) => <Checkbox label="Default Checkbox" {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: "Default Checkbox",
  // eslint-disable-next-line no-console
  onChange: console.log,
};

export const NoLabel = Template.bind({});
NoLabel.args = {
  label: undefined,
};

export const Checked = Template.bind({});
Checked.args = {
  label: "Checked Checkbox",
  checked: true,
};

export const SemiChecked = Template.bind({});
SemiChecked.args = {
  label: "Semi Checked Checkbox",
  semi: true,
};

export const Round = Template.bind({});
Round.args = {
  label: "Round Checkbox",
  round: true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  label: "Disabled Checkbox",
  disabled: true,
};

export const HTMLName = Template.bind({});
HTMLName.args = {
  label: "HTMLName Checkbox",
  name: "test-name",
};

export const Variants = () => {
  const c = <Checkbox checked />;
  return (
    <>
      {[
        "primary",
        "secondary",
        "tertiary",
        "success",
        "danger",
        "warning",
        "dark",
        "light",
      ].map((kind) =>
        React.cloneElement(c, {
          ...c.props,
          kind,
          label: kind,
          name: kind,
          disabled: true,
        })
      )}
    </>
  );
};
