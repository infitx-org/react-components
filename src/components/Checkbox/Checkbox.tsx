import classnames from "classnames";
import React, { ChangeEvent } from "react";
import { Kind } from "types";
import "./Checkbox.scss";

export interface CheckboxProps {
  kind: `${Kind}`;
  style: React.CSSProperties;
  className: string;
  label: string;
  name?: string;
  id?: string;
  checked?: boolean;
  semi?: boolean;
  round?: boolean;
  disabled?: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
}

function Checkbox({
  kind = "primary",
  style,
  className,
  label,
  name,
  id,
  checked = false,
  semi = false,
  round = false,
  disabled = false,
  onChange,
}: CheckboxProps) {
  const [isChecked, setChecked] = React.useState(checked);
  React.useEffect(() => {
    setChecked(checked);
  }, [checked]);

  const wrapperClassName = classnames([
    "mb-input",
    "rc-checkbox__wrapper",
    className,
  ]);
  const checkboxClassName = classnames([
    "rc-checkbox",
    `rc-checkbox--${kind}`,
    semi && "rc-checkbox--semi-checked",
    round && "rc-checkbox--round",
    !label && "rc-checkbox--no-margin",
  ]);
  const labelClassName = classnames([
    "rc-checkbox__label",
    disabled && "rc-checkbox__label--disabled",
  ]);
  return (
    <label className={wrapperClassName} style={style}>
      <input
        type="checkbox"
        id={id}
        name={name}
        className={checkboxClassName}
        onChange={(e) => {
          setChecked(!isChecked);
          onChange(e);
        }}
        checked={isChecked && semi !== true}
        disabled={disabled}
      />
      {label && <span className={labelClassName}>{label}</span>}
    </label>
  );
}

export default React.memo(Checkbox);
