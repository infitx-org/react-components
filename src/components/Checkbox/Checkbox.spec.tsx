import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import Checkbox from "./Checkbox";

const commonProps = {
  name: "test-name",
  onChange: jest.fn(),
  label: "test-label",
};

describe("tests the checbox", () => {
  it("renders the checkbox group", () => {
    const { container } = render(<Checkbox {...commonProps} />);
    expect(container.querySelector(".rc-checkbox__wrapper")).toBeTruthy();
    expect(
      container.querySelector('input[type="checkbox"]')
    ).toBeInTheDocument();
  });

  it("renders the label prop", () => {
    const { container } = render(<Checkbox {...commonProps} />);
    expect(container.querySelector(".rc-checkbox__label")).toHaveTextContent(
      "test-label"
    );
  });

  it("renders the id prop", () => {
    const { container } = render(<Checkbox {...commonProps} id="test-id" />);
    expect(container.querySelector('input[type="checkbox"]')).toHaveAttribute(
      "id",
      "test-id"
    );
  });

  it("renders the name prop", () => {
    const { container } = render(<Checkbox {...commonProps} />);
    expect(container.querySelector('input[type="checkbox"]')).toHaveAttribute(
      "name",
      "test-name"
    );
  });

  it("renders the checked prop", () => {
    const { container } = render(<Checkbox {...commonProps} checked />);
    expect(
      container.querySelector('input[type="checkbox"][checked]')
    ).toBeInTheDocument();
  });

  it("renders the disabled prop", () => {
    const { container } = render(<Checkbox {...commonProps} disabled />);
    expect(
      container.querySelector('input[type="checkbox"][disabled]')
    ).toBeInTheDocument();
  });

  it("renders the semi prop", () => {
    const { container } = render(<Checkbox {...commonProps} semi />);
    expect(
      container.querySelector(
        'input[type="checkbox"].rc-checkbox--semi-checked'
      )
    ).toBeInTheDocument();
  });

  it("renders the round prop", () => {
    const { container } = render(<Checkbox {...commonProps} round />);
    expect(
      container.querySelector('input[type="checkbox"].rc-checkbox--round')
    ).toBeInTheDocument();
  });

  it("triggers the onChange", () => {
    const mockEvent = jest.fn();
    const { container } = render(
      <Checkbox {...commonProps} onChange={mockEvent} />
    );
    userEvent.click(
      container.querySelector('input[type="checkbox"]') as Element
    );
    expect(mockEvent).toHaveBeenCalledWith(true);
    // TODO: Let's see if using boolean is better than changeEvent
    // expect.objectContaining({
    //   type: "change",
    //   target: expect.objectContaining({ checked: true }),
    // })
    // );
  });
});

// Snapshot testing
it("renders the Checkbox correctly when multiple props are set", () => {
  const { container } = render(<Checkbox {...commonProps} />);
  expect(container).toMatchSnapshot();
});
