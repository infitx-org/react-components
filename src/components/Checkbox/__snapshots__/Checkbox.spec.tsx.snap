// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Checkbox correctly when multiple props are set 1`] = `
<div>
  <label
    class="mb-input input-checkbox__wrapper"
  >
    <input
      class="input-checkbox"
      name="test-name"
      type="checkbox"
    />
    <span
      class="input-checkbox__label"
    >
      test-label
    </span>
  </label>
</div>
`;
